### DISCUSSION TITLE
using-Statement with multiple Disposables of different types

### DISCUSSION THREAD

**Discussion by bspinner** [dotnet/csharplang#1579] [Language Ideas]
There are moments we tend to write constructs like

```cs
using(await AsyncLock())
{
  using(var y = await WhateverElseAsync())
  {
    using(var x = await WhateverAsync())
    {
      // do stuff
    }
  }
}
```

or

```cs
using(await AsyncLock())
using(var y = await WhateverElseAsync())
using(var x = await WhateverAsync())
{
      // do stuff
}
```

Unfortunately one can't combine those using Statements into one as long as the type isn't the same for all disposables.

My proposal is to enable developers to combine all those using statements into one.

```cs
using(await AsyncLock(); var y = await WhateverElseAsync(); var x = await WhateverAsync())
{
      // do stuff
}
```

**Comment by ufcpp** [comment]
related to https://github.com/dotnet/csharplang/issues/1473

**Comment by qrli** [comment]
A bit off-topic: what's the point of using lock in async?

**Comment by jnm2** [comment]
@qrli An async lock is basically a queue, isn't it?

**Comment by franchyd** [comment]
What is wrong with the second example you wrote there?  I'm not sure how using a single block is better?

**Comment by bbarry** [comment]
@qrli: [off topic] The point of locking is to control access to some resource. You cannot use `await` inside a lock, so code that would need synchronization needs another design. For example serializing access to a write buffer that you want to write to using an async write.

Conceptually `AsyncLock` fills a queue of task completion sources when the lock is under contention and dequeues one and marks it complete when the previous user of the lock releases hold.

https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncLock.cs#L100

edit for anyone looking in the future: while the implementation in AsyncEx is relatively easy to understand it is almost certainly heavier than @stephentoub's implementation:

    public sealed class AsyncLock
    {
        private readonly SemaphoreSlim m_semaphore = new SemaphoreSlim(1, 1);
        private readonly Task<IDisposable> m_releaser;
    
        public AsyncLock()
        {
            m_releaser = Task.FromResult((IDisposable)new Releaser(this));
        }
    
        public Task<IDisposable> LockAsync()
        {
            var wait = m_semaphore.WaitAsync();
            return wait.IsCompleted ?
                        m_releaser :
                        wait.ContinueWith((_, state) => (IDisposable)state,
                            m_releaser.Result, CancellationToken.None,
            TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
        }
    
        private sealed class Releaser : IDisposable
        {
            private readonly AsyncLock m_toRelease;
            internal Releaser(AsyncLock toRelease) { m_toRelease = toRelease; }
            public void Dispose() { m_toRelease.m_semaphore.Release(); }
        }
    }

As neither implementations are [reentrant](https://wikipedia.org/wiki/Reentrant_mutex), I am inclined to believe that this smaller implementation is better (not convinced either that a reentrant solution is desirable anyway).

**Comment by bbarry** [comment]
Back on topic, I don't like this.

Just a gut reaction but I don't see the point of saving a few keystrokes this way. I'd much rather see #93 happen and then you could get something similar to this via a tuple syntax. #93 also solves the problem of ref structs not being able to be disposable.

**Comment by bspinner** [comment]
@DanFTRX I dislike that for the same reasons using single line if statements without braces. :)

@bbarry It's not just about saving few keystrokes. In my opionion a block has a higher readability and is less error prone than using `using` without braces.

**Comment by bbarry** [comment]
I don't like seeing using statements without braces either.

To be clear my issue here is with multiple expression statements inside of a using statement, which creates a context around a statement that changes its behavior. I like the idea of improving how using statements work (in particular #93 and #114). I do not like messing with context of expression statements.

**Comment by Neme12** [comment]
What's wrong with this?
```c#
using(await AsyncLock())
using(var y = await WhateverElseAsync())
using(var x = await WhateverAsync())
{
      // do stuff
}
```
I don't see any problem here.

**Comment by jnm2** [comment]
Besides the missing spaces after the after the `using` keywords...
